Allow cjson.replace variables to be Objects or Arrays

rafzi opened the pull request:
This allows for great composability of config files.

```
base.cjson:
{
    "opt": 3,
    "ext": {{ext}}
}

ext.cjson:
{
    "ext": {
        "more": 5
    }
}
```

Previously this was only possible by making a variable a single line string, which is not useful for big blocks.

First I went for cjson.extend to manually add sub-objects where they belong, but this quickly got messy and unclear.

Related: What do you think about an operation similar to "merge" for replace, where you give it a base and ordered replacements? It would replace and extend from last to first file to gather all variables into a final file.

kof commented on index.js:
```diff
@@ -99,7 +99,10 @@ exports.parse = function(str, reviver) {
  */
 exports.replace = function(str, data) {
     return str.replace(/\{\{([^}]+)\}\}/g, function(match, search) {
-        return data[search] ? data[search] : match;
```
I think ternary inside of a ternary is a bad readability.

kof commented on index.js:
```diff
@@ -99,7 +99,10 @@ exports.parse = function(str, reviver) {
-        return data[search] ? data[search] : match;
+        // if the variable is an object, it needs to be stringified first
+        return data[search] ?
+            data[search] === Object(data[search]) ? JSON.stringify(data[search]) : data[search]
```
why not to use typeof?

kof commented:
Can you please also add a test?

rafzi commented:
I updated the PR with your feedback.

I only know the basics of javascript, so I used this as my source for the object detection:
http://stackoverflow.com/a/22482737/2040925

Using `typeof` would detect `null` as an object, but since that does not matter here I changed the code to use typeof.

The changes now also permit replacements that evaluated as false like `null`, `false` and `""`.

kof commented:
Olegs-MBP:node-cjson kof [master] $ npm run test

> cjson@0.3.3 test /Users/kof/work/projects/kof/node-cjson
> node ./test/test.js


assert.js:89
  throw new assert.AssertionError({
  ^
AssertionError: JSON types as template variables
    at Object.<anonymous> (/Users/kof/work/projects/kof/node-cjson/test/test.js:64:3)
    at Module._compile (module.js:434:26)
    at Object.Module._extensions..js (module.js:452:10)
    at Module.load (module.js:355:32)
    at Function.Module._load (module.js:310:12)
    at Function.Module.runMain (module.js:475:10)
    at startup (node.js:117:18)
    at node.js:951:3

kof commented:
When you contribute to a project, try to make your code look like the rest of it. It  includes naming conventions, file names, quotes etc.

kof commented on index.js:
```diff
@@ -99,7 +99,17 @@ exports.parse = function(str, reviver) {
 exports.replace = function(str, data) {
     return str.replace(/\{\{([^}]+)\}\}/g, function(match, search) {
-        return data[search] ? data[search] : match;
+        if (data.hasOwnProperty(search)) {
```
not necessary, data[search] is ok as we assume plain objects here.

rafzi replied:
This did fail for me when the replacement variable is `false` for example:

```
{
    "debug": {{debug}}
}

{
    "debug": false
}
```

kof replied:
in this case if (data[search] != null)

rafzi replied:
That would exclude `null` as possible variable, similarily to `false`. See the second test case.

kof replied:
I see, then its fine.

kof commented on index.js:
```diff
@@ -99,7 +99,17 @@ exports.parse = function(str, reviver) {
+        if (data.hasOwnProperty(search)) {
+            // If the variable is an object, stringify it before replacement.
+            // The false positive of "null" is fine in this case.
+            if (typeof data[search] === "object") {
```
have you seen double quotes in code anywhere?

kof commented on test/fixtures/conf_objvar_base.json:
```diff
@@ -0,0 +1,4 @@
+{
```
how about to have consistent fixture file names?

rafzi commented:
Sorry for the failing test. I have zero knowledge about node and only did the testing manually. I didn't save a file.

What exactly do you want me to change convention wise? I don't have that much code to look at. Do you use camel case?

I gave the conf files useful names, so they are easier to differentiate as more are added. I could rename the rest in an unrelated PR for consistency.

kof commented on index.js:
```diff
@@ -99,7 +99,17 @@ exports.parse = function(str, reviver) {
+            // The false positive of "null" is fine in this case.
+            if (typeof data[search] === 'object') {
+                return JSON.stringify(data[search]);
+            } else {
```
else is not needed, you did an early return before

kof commented on index.js:
```diff
@@ -99,7 +99,17 @@ exports.parse = function(str, reviver) {
+            } else {
+                return data[search];
+            }
+        } else {
```
same here ...

kof commented:
still getting the error


 cjson@0.3.3 test /Users/kof/work/projects/kof/node-cjson
> node ./test/test.js

/Users/kof/work/projects/kof/node-cjson/index.js:234
        throw err;
        ^

SyntaxError: Unexpected token 's' at 2:20
    "subobject": {{subobject}},
                   ^
File: "/Users/kof/work/projects/kof/node-cjson/test/fixtures/conf12.json"
    at SyntaxError (native)
    at Object.parseJSON (/Users/kof/work/projects/kof/node-cjson/node_modules/json-parse-helpfulerror/node_modules/jju/lib/parse.js:732:13)
    at Object.parse (/Users/kof/work/projects/kof/node-cjson/node_modules/json-parse-helpfulerror/index.js:10:13)
    at Object.exports.parse (/Users/kof/work/projects/kof/node-cjson/index.js:90:28)
    at load (/Users/kof/work/projects/kof/node-cjson/index.js:231:24)
    at /Users/kof/work/projects/kof/node-cjson/index.js:194:24
    at Array.forEach (native)
    at load (/Users/kof/work/projects/kof/node-cjson/index.js:193:14)
    at Object.load (/Users/kof/work/projects/kof/node-cjson/index.js:218:16)
    at Object.<anonymous> (/Users/kof/work/projects/kof/node-cjson/test/test.js:83:19)

rafzi commented:
I greatly overestimated the complexity of running tests on node. Was easy and painless.

Should be all good now:

```
user1@debj:~/tmp_cjson/node-cjson$ npm run test

> cjson@0.3.3 test /home/user1/tmp_cjson/node-cjson
> node ./test/test.js

All tests passed.

```

kof commented:
merged and published, thanks
